public/app/features/alerting/unified/rule-list/hooks/datasourceFilter.ts

```diff
@@ -22,6 +22,27 @@ import {
   ruleTypeFilter,
 } from './filterPredicates';
 
+/**
+ * Determines if client-side filtering is needed for data source-managed rules.
+ */
+export function hasDatasourceFilters(filterState: Partial<RulesFilter>): boolean {
```

Comment:
nit: this function name seems confusing / unrelated to data sources being included in the filter :)

Perhaps `shouldPerformClientSideFiltering()` or similar ?